packages/core/js-peer/build.ts

```diff
@@ -0,0 +1,103 @@
+import path, { dirname } from 'path';
+import type { InlineConfig, PluginOption } from 'vite';
+import { build } from 'vite';
+import { builtinModules, createRequire } from 'module';
+import tsconfigPaths from 'vite-tsconfig-paths';
+import inject from '@rollup/plugin-inject';
+import stdLibBrowser from 'node-stdlib-browser';
+import { fileURLToPath } from 'url';
+import { replaceCodePlugin } from 'vite-plugin-replace';
+import pkg from './package.json' assert { type: 'json' };
+
+const require = createRequire(import.meta.url);
+
+
+const commonConfig = (isNode: boolean): InlineConfig & Required<Pick<InlineConfig, 'build'>> => {
+    const esbuildShim = require.resolve('node-stdlib-browser/helpers/esbuild/shim');
+    return {
+        build: {
+            target: isNode ? 'es2022' : 'modules',
+            minify: 'esbuild',
+            lib: {
+                entry: './src/index.ts',
+                name: 'js-client',
+                fileName: 'index',
+            },
+            outDir: isNode ? './dist/node' : './dist/browser',
+            ...(isNode ? {
+                rollupOptions: {
+                    external: [...builtinModules, ...builtinModules.map(bm => `node:${bm}`)],
+                    plugins: [
+                        // @ts-ignore
+                        inject({
+                            self: 'global',
+                            'WorkerScope': ['worker_threads', '*'],
+                            'Worker': ['worker_threads', 'Worker'],
+                            'isMainThread': ['worker_threads', 'isMainThread'],
+                        })
+                    ]
+                }
+            } : {
+                rollupOptions: {
+                    plugins: [
+                        {
+                            // @ts-ignore
+                            ...inject({
+                                global: [esbuildShim, 'global'],
+                                process: [esbuildShim, 'process'],
+                                Buffer: [esbuildShim, 'Buffer']
+                            }), enforce: 'post'
+                        }
+                    ]
+                }
+            })
+        },
+        plugins: [tsconfigPaths(), ...(isNode ? [replaceCodePlugin({
+            replacements: [
+                { from: 'require(`./${file}.js`)', to: 'require(`./linux.js`)' },
+                {
+                    from: 'const { name, version } = req(\'../../package.json\')',
+                    to: 'const { name, version } = { name: \'ssdp\', version: \'4.0.4\' }'
+                },
+                { from: 'eval("require")("worker_threads")', to: 'WorkerScope' },
```

Comment:
could you please help me understand how this is working? I understand that this should replace the following strings before bundling but I don't see these string anywhere? Where do they come from?

Reply:
This is related to @achingbrain/* package family, which is not properly adapted to different environments. Probably they are node specific. Nevertheless, libp2p brings them to main bundle so it renders bundle unusable because of these remained CommonJS imports.